modules/videoio/src/cap_gphoto2.cpp

```diff
@@ -0,0 +1,1218 @@
+        context = NULL;
+    }
+    catch (GPhoto2Exception & e)
+    {
+        message(ERROR, "destruction error", e);
+    }
+}
+
+/**
+ * Connects to selected device.
+ */
+bool DigitalCameraCapture::open(int index)
+{
+    const char * model = 0, *path = 0;
+    int m, p;
+    GPPortInfo portInfo;
+
+    if (isOpened()) {
+        close();
+    }
+
+    try
+    {
+        CR(gp_camera_new(&camera));
+        CR(gp_list_get_name(allDevices, index, &model));
+        CR(gp_list_get_value(allDevices, index, &path));
+
+        // Set model abilities.
+        CR(m = gp_abilities_list_lookup_model(abilitiesList, model));
+        CR(gp_abilities_list_get_abilities(abilitiesList, m, &cameraAbilities));
+        CR(gp_camera_set_abilities(camera, cameraAbilities));
+
+        // Set port
+        CR(p = gp_port_info_list_lookup_path(capablePorts, path));
+        CR(gp_port_info_list_get_info(capablePorts, p, &portInfo));
+        CR(gp_camera_set_port_info(camera, portInfo));
+
+        // Initialize connection to the camera.
+        CR(gp_camera_init(camera, context));
+
+        message(STATUS, "connected camera", model);
+        message(STATUS, "connected using", path);
+
+        // State initialization
+        firstCapturedFrameTime = 0;
+        capturedFrames = 0;
+        preview = false;
+        reloadOnChange = false;
+        collectMsgs = false;
+
+        reloadConfig();
+
+        opened = true;
+        return true;
+    }
+    catch (GPhoto2Exception & e)
+    {
+        message(WARNING, "opening device failed", e);
+        return false;
+    }
+}
+
+/**
+ *
+ */
+bool DigitalCameraCapture::isOpened() const
+{
+    return opened;
+}
+
+/**
+ * Close connection to the camera. Remove all unread frames/files.
+ */
+void DigitalCameraCapture::close()
+{
+    try
+    {
+        if (!frame.empty())
+        {
+            frame.release();
+        }
+        if (camera)
+        {
+            CR(gp_camera_exit(camera, context));
+            CR(gp_camera_unref(camera));
+            camera = NULL;
+        }
+        opened = false;
+        if (int frames = grabbedFrames.size() > 0)
+        {
+            while (frames--)
+            {
+                CameraFile * file = grabbedFrames.front();
+                grabbedFrames.pop_front();
+                CR(gp_file_unref(file));
+            }
+        }
+        if (rootWidget)
+        {
+            widgetInfo.clear();
+            CR(gp_widget_unref(rootWidget));
+            rootWidget = NULL;
+        }
+    }
+    catch (GPhoto2Exception & e)
+    {
+        message(ERROR, "cannot close device properly", e);
+    }
+}
+
+/**
+ * @param output will be changed if possible, return 0 if changed,
+ * @return widget, or NULL if output value was found (saved in argument),
+ */
+CameraWidget * DigitalCameraCapture::getGenericProperty(int propertyId,
+        double & output) const
+{
+    switch (propertyId)
+    {
+    case CV_CAP_PROP_POS_MSEC:
```

Comment:
The cases below do not have a correct identation. They should be 1 further than the switch statement. Could you fix?